File: apps/api/src/routes/lastfm.ts
import { Hono } from 'hono';
import { prisma, getSettings } from '../store.js';
import { requireCredentials, sign, signedPost, signedGet, sessionKey } from '../lastfm.js';

const lastfm = new Hono();

lastfm.get('/auth', async (c) => {
  const settings = await getSettings();
  try {
    const { apiKey, secret } = requireCredentials(settings);
    const params = { method: 'auth.getToken', api_key: apiKey };
    const sig = sign(params, secret);
    const url = new URL('https://ws.audioscrobbler.com/2.0/');
    for (const [k, v] of Object.entries({ ...params, api_sig: sig, format: 'json' })) {
      url.searchParams.set(k, v);
    }
    const res = await fetch(url);
    const json = (await res.json()) as Record<string, unknown> & { error?: number; message?: string; token?: string };
    if (json.error) throw new Error(json.message ?? 'Failed to get auth token');
    return c.json({
      token: String(json.token),
      authUrl: `https://www.last.fm/api/auth/?api_key=${apiKey}&token=${json.token}`
    });
  } catch (error) {
    return c.json({ error: error instanceof Error ? error.message : 'Failed to start Last.fm auth' }, 502);
  }
});

lastfm.post('/auth', async (c) => {
  const settings = await getSettings();
  try {
    const body = await c.req.json<Record<string, unknown>>();
    const token = String(body?.token ?? '').trim();
    if (!token) return c.json({ error: 'token is required' }, 400);
    const json = await signedPost(settings, { method: 'auth.getSession', token });
    const session = json.session as Record<string, string>;
    const username = String(session?.name ?? '');
    const sk = String(session?.key ?? '');
    if (!sk) throw new Error('No session key returned from Last.fm');
    await prisma.$transaction([
      prisma.setting.upsert({ where: { key: 'LASTFM_SESSION_KEY' }, update: { value: sk }, create: { key: 'LASTFM_SESSION_KEY', value: sk } }),
      prisma.setting.upsert({ where: { key: 'LASTFM_USERNAME' }, update: { value: username }, create: { key: 'LASTFM_USERNAME', value: username } })
    ]);
    return c.json({ username });
  } catch (error) {
    return c.json({ error: error instanceof Error ? error.message : 'Failed to complete Last.fm auth' }, 502);
  }
});

lastfm.post('/disconnect', async (c) => {
  await prisma.$transaction([
    prisma.setting.upsert({ where: { key: 'LASTFM_SESSION_KEY' }, update: { value: '' }, create: { key: 'LASTFM_SESSION_KEY', value: '' } }),
    prisma.setting.upsert({ where: { key: 'LASTFM_USERNAME' }, update: { value: '' }, create: { key: 'LASTFM_USERNAME', value: '' } })
  ]);
  return c.json({ ok: true });
});

lastfm.post('/now-playing', async (c) => {
  const settings = await getSettings();
  try {
    const body = await c.req.json<Record<string, unknown>>();
    const artist = String(body?.artist ?? '').trim();
    const track = String(body?.track ?? '').trim();
    if (!artist || !track) return c.json({ error: 'artist and track are required' }, 400);
    const sk = sessionKey(settings);
    if (sk) {
      const params: Record<string, string> = { method: 'track.updateNowPlaying', sk, artist, track };
      if (body?.album) params.album = String(body.album);
      if (body?.duration) params.duration = String(Math.round(Number(body.duration)));
      await signedPost(settings, params);
    }
    return c.json({ ok: true });
  } catch (error) {
    return c.json({ error: error instanceof Error ? error.message : 'Now playing update failed' }, 502);
  }
});

lastfm.post('/scrobble', async (c) => {
  const settings = await getSettings();
  try {
    const body = await c.req.json<Record<string, unknown>>();
    const artist = String(body?.artist ?? '').trim();
    const track = String(body?.track ?? '').trim();
    const timestamp = Number(body?.timestamp);
    if (!artist || !track || Number.isNaN(timestamp)) {
      return c.json({ error: 'artist, track, and timestamp are required' }, 400);
    }
    const sk = sessionKey(settings);
    if (sk) {
      const params: Record<string, string> = {
        method: 'track.scrobble',
        sk,
        artist,
        track,
        timestamp: String(Math.floor(timestamp))
      };
      if (body?.album) params.album = String(body.album);
      if (body?.duration) params.duration = String(Math.round(Number(body.duration)));
      await signedPost(settings, params);
    }
    return c.json({ ok: true });
  } catch (error) {
    return c.json({ error: error instanceof Error ? error.message : 'Scrobble failed' }, 502);
  }
});

lastfm.get('/user-taste', async (c) => {
  const settings = await getSettings();
  const sk = sessionKey(settings);
  const username = String(settings.LASTFM_USERNAME ?? '');
  if (!sk) return c.json({ connected: false, username, artists: [] });
  try {
    const json = await signedGet(settings, { method: 'user.getTopArtists', sk, period: '3month', limit: '50' });
    const topartists = json?.topartists as Record<string, unknown> | undefined;
    const artists = Array.isArray(topartists?.artist)
      ? (topartists.artist as Record<string, unknown>[]).map((a) => String(a?.name ?? '')).filter(Boolean)
      : [];
    return c.json({ connected: true, username, artists });
  } catch {
    return c.json({ connected: true, username, artists: [] });
  }
});

export default lastfm;

File: src-tauri/src/commands/lastfm.rs
use std::collections::HashMap;
use serde::Serialize;
use tauri::State;
use crate::{AppState, commands::settings};

fn get_lfm_tokens(state: &AppState) -> (String, String, String) {
    let Ok(db) = state.db.lock() else { return (String::new(), String::new(), String::new()) };
    let Ok(s) = settings::read_all(&db) else { return (String::new(), String::new(), String::new()) };
    (
        s.get("LASTFM_API_KEY").cloned().unwrap_or_default(),
        s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default(),
        s.get("LASTFM_SESSION_KEY").cloned().unwrap_or_default(),
    )
}

const LASTFM_API: &str = "https://ws.audioscrobbler.com/2.0/";

// ── Signing ──────────────────────────────────────────────────────────────────

fn sign(params: &HashMap<String, String>, secret: &str) -> String {
    let mut pairs: Vec<(&String, &String)> = params
        .iter()
        .filter(|(k, _)| *k != "format" && *k != "callback")
        .collect();
    pairs.sort_by_key(|(k, _)| k.as_str());
    let mut s = String::new();
    for (k, v) in pairs {
        s.push_str(k);
        s.push_str(v);
    }
    s.push_str(secret);
    format!("{:x}", md5::compute(s.as_bytes()))
}

/// Public so `subsonic.rs` can call it for track.love / track.unlove.
pub async fn sign_and_post(
    http: &reqwest::Client,
    mut params: HashMap<String, String>,
    api_key: &str,
    secret: &str,
) -> Result<serde_json::Value, String> {
    params.insert("api_key".to_string(), api_key.to_string());
    let sig = sign(&params, secret);
    params.insert("api_sig".to_string(), sig);
    params.insert("format".to_string(), "json".to_string());

    let resp = http
        .post(LASTFM_API)
        .form(&params)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    let json: serde_json::Value = resp.json().await.map_err(|e| e.to_string())?;
    if let Some(code) = json.get("error") {
        let msg = json.get("message").and_then(|m| m.as_str()).unwrap_or("Last.fm error");
        return Err(format!("Last.fm error {code}: {msg}"));
    }
    Ok(json)
}

async fn sign_and_get(
    http: &reqwest::Client,
    mut params: HashMap<String, String>,
    api_key: &str,
    secret: &str,
) -> Result<serde_json::Value, String> {
    params.insert("api_key".to_string(), api_key.to_string());
    let sig = sign(&params, secret);
    params.insert("api_sig".to_string(), sig);
    params.insert("format".to_string(), "json".to_string());

    let resp = http.get(LASTFM_API).query(&params).send().await.map_err(|e| e.to_string())?;
    let json: serde_json::Value = resp.json().await.map_err(|e| e.to_string())?;
    if let Some(code) = json.get("error") {
        let msg = json.get("message").and_then(|m| m.as_str()).unwrap_or("Last.fm error");
        return Err(format!("Last.fm error {code}: {msg}"));
    }
    Ok(json)
}

fn require_credentials(s: &HashMap<String, String>) -> Result<(String, String), String> {
    let key = s.get("LASTFM_API_KEY").cloned().unwrap_or_default();
    let secret = s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default();
    if key.is_empty() || secret.is_empty() {
        return Err("Last.fm API key and shared secret are required".to_string());
    }
    Ok((key, secret))
}

fn read_session_credentials(s: &HashMap<String, String>) -> (String, String, String) {
    let key = s.get("LASTFM_API_KEY").cloned().unwrap_or_default();
    let secret = s.get("LASTFM_SHARED_SECRET").cloned().unwrap_or_default();
    let sk = s.get("LASTFM_SESSION_KEY").cloned().unwrap_or_default();
    (key, secret, sk)
}

// ── Commands ─────────────────────────────────────────────────────────────────

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LfmAuthToken {
    pub token: String,
    pub auth_url: String,
}

#[tauri::command]
pub async fn lfm_begin_auth(state: State<'_, AppState>) -> Result<LfmAuthToken, String> {
    let (api_key, secret) = {
        let (k, s, _) = get_lfm_tokens(&state);
        if !k.is_empty() {
            (k, s)
        } else {
            let db = state.db.lock().map_err(|e| e.to_string())?;
            require_credentials(&settings::read_all(&db)?)?
        }
    };
    let mut params = HashMap::new();
    params.insert("method".to_string(), "auth.getToken".to_string());
    let json = sign_and_get(&state.http, params, &api_key, &secret).await?;
    let token = json
        .get("token")
        .and_then(|t| t.as_str())
        .ok_or_else(|| "Failed to get auth token".to_string())?
        .to_string();
    Ok(LfmAuthToken {
        auth_url: format!("https://www.last.fm/api/auth/?api_key={api_key}&token={token}"),
        token,
    })
}

#[derive(Serialize)]
pub struct LfmSession {
    pub username: String,
}

#[tauri::command]
pub async fn lfm_complete_auth(
    state: State<'_, AppState>,
    token: String,
) -> Result<LfmSession, String> {
    let (api_key, secret) = {
        let (k, s, _) = get_lfm_tokens(&state);
        if !k.is_empty() {
            (k, s)
        } else {
            let db = state.db.lock().map_err(|e| e.to_string())?;
            require_credentials(&settings::read_all(&db)?)?
        }
    };
    let mut params = HashMap::new();
    params.insert("method".to_string(), "auth.getSession".to_string());
    params.insert("token".to_string(), token.trim().to_string());
    let json = sign_and_get(&state.http, params, &api_key, &secret).await?;
    let session = json.get("session").ok_or_else(|| "No session in Last.fm response".to_string())?;
    let username = session.get("name").and_then(|n| n.as_str()).unwrap_or("").to_string();
    let sk = session.get("key").and_then(|k| k.as_str()).unwrap_or("").to_string();
    if sk.is_empty() {
        return Err("No session key returned from Last.fm".to_string());
    }
    let db = state.db.lock().map_err(|e| e.to_string())?;
    settings::upsert(&db, "LASTFM_SESSION_KEY", &sk)?;
    settings::upsert(&db, "LASTFM_USERNAME", &username)?;
    Ok(LfmSession { username })
}

#[tauri::command]
pub fn lfm_disconnect(state: State<'_, AppState>) -> Result<(), String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    settings::upsert(&db, "LASTFM_SESSION_KEY", "")?;
    settings::upsert(&db, "LASTFM_USERNAME", "")?;
    Ok(())
}

#[tauri::command]
pub async fn lfm_now_playing(
    state: State<'_, AppState>,
    artist: String,
    track: String,
    album: Option<String>,
    duration: Option<f64>,
) -> Result<(), String> {
    let (api_key, secret, sk) = {
        let tokens = get_lfm_tokens(&state);
        if !tokens.0.is_empty() {
            tokens
        } else {
            let db = state.db.lock().map_err(|e| e.to_string())?;
            read_session_credentials(&settings::read_all(&db)?)
        }
    };
    if sk.is_empty() || api_key.is_empty() || secret.is_empty() {
        return Ok(());
    }
    let mut params = HashMap::new();
    params.insert("method".to_string(), "track.updateNowPlaying".to_string());
    params.insert("artist".to_string(), artist.trim().to_string());
    params.insert("track".to_string(), track.trim().to_string());
    if let Some(album) = album.filter(|a| !a.trim().is_empty()) {
        params.insert("album".to_string(), album.trim().to_string());
    }
    if let Some(dur) = duration.filter(|d| *d > 0.0) {
        params.insert("duration".to_string(), (dur as u64).to_string());
    }
    params.insert("sk".to_string(), sk);
    let _ = sign_and_post(&state.http, params, &api_key, &secret).await;
    Ok(())
}

#[tauri::command]
pub async fn lfm_scrobble(
    state: State<'_, AppState>,
    artist: String,
    track: String,
    timestamp: i64,
    album: Option<String>,
    duration: Option<f64>,
) -> Result<(), String> {
    let (api_key, secret, sk) = {
        let tokens = get_lfm_tokens(&state);
        if !tokens.0.is_empty() {
            tokens
        } else {
            let db = state.db.lock().map_err(|e| e.to_string())?;
            read_session_credentials(&settings::read_all(&db)?)
        }
    };
    if sk.is_empty() || api_key.is_empty() || secret.is_empty() {
        return Ok(());
    }
    let mut params = HashMap::new();
    params.insert("method".to_string(), "track.scrobble".to_string());
    params.insert("artist".to_string(), artist.trim().to_string());
    params.insert("track".to_string(), track.trim().to_string());
    params.insert("timestamp".to_string(), timestamp.to_string());
    if let Some(album) = album.filter(|a| !a.trim().is_empty()) {
        params.insert("album".to_string(), album.trim().to_string());
    }
    if let Some(dur) = duration.filter(|d| *d > 0.0) {
        params.insert("duration".to_string(), (dur as u64).to_string());
    }
    params.insert("sk".to_string(), sk);
    let _ = sign_and_post(&state.http, params, &api_key, &secret).await;
    Ok(())
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UserTaste {
    pub connected: bool,
    pub username: String,
    pub artists: Vec<String>,
}

#[tauri::command]
pub async fn lfm_user_taste(state: State<'_, AppState>) -> Result<UserTaste, String> {
    let (api_key, _secret, sk, username) = {
        let (k, sec, sk) = get_lfm_tokens(&state);
        let db = state.db.lock().map_err(|e| e.to_string())?;
        let s = settings::read_all(&db)?;
        let u = s.get("LASTFM_USERNAME").cloned().unwrap_or_default();
        if !k.is_empty() {
            (k, sec, sk, u)
        } else {
            let (k2, sec2, sk2) = read_session_credentials(&s);
            (k2, sec2, sk2, u)
        }
    };
    if sk.is_empty() || api_key.is_empty() {
        return Ok(UserTaste { connected: false, username, artists: vec![] });
    }
    let result = state.http
        .get(LASTFM_API)
        .query(&[
            ("method", "user.getTopTracks"),
            ("user", username.as_str()),
            ("period", "3month"),
            ("limit", "100"),
            ("api_key", api_key.as_str()),
            ("format", "json"),
        ])
        .send()
        .await
        .map_err(|e| e.to_string())?
        .json::<serde_json::Value>()
        .await
        .map_err(|e| e.to_string());

    match result {
        Ok(json) => {
            let tracks = json
                .get("toptracks")
                .and_then(|tt| tt.get("track"))
                .and_then(|t| t.as_array())
                .cloned()
                .unwrap_or_default();
            let mut artists: Vec<String> = Vec::new();
            for t in tracks {
                let artist = t.get("artist")
                    .and_then(|a| a.get("name"))
                    .and_then(|n| n.as_str())
                    .unwrap_or("")
                    .trim()
                    .to_string();
                if !artist.is_empty() && !artists.contains(&artist) {
                    artists.push(artist);
                }
                if artists.len() >= 50 {
                    break;
                }
            }
            Ok(UserTaste { connected: true, username, artists })
        }
        Err(_) => Ok(UserTaste { connected: true, username, artists: vec![] }),
    }
}

#[derive(Serialize)]
pub struct LfmStatus {
    pub connected: bool,
    pub username: String,
}

#[tauri::command]
pub fn lfm_status(state: State<'_, AppState>) -> Result<LfmStatus, String> {
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let s = settings::read_all(&db)?;
    let connected = s.get("LASTFM_SESSION_KEY").map(|v| !v.is_empty()).unwrap_or(false);
    let username = s.get("LASTFM_USERNAME").cloned().unwrap_or_default();
    Ok(LfmStatus { connected, username })
}
